### ISSUE TITLE
expect_loop could eat all memory?

### ISSUE THREAD

**Issue by CKD4XXH** [pexpect/pexpect#388]
I have not tested it, but the (master) code looks as if the expect_loop function is doing essentially a `while timeout_not_reached: spawn.buffer += spawn.read_nonblocking(...)` operation.
So it seems to be possible that `read_nonblocking` always return in a few milliseconds with new megabytes of data, which then gets appended to the buffer. Without upper limit. Maybe for hours or days.
What happens when the RAM fills up?..

**Comment by takluyver** [comment]
IIRC, it will stop adding to the buffer when a pattern you're looking for
matches. So if a lot of data comes without matching the pattern, then yes,
it can fill up your memory. But pexpect is often used to find a pattern and
then see what came before that, so we can't easily discard that data. If
you're likely to be dealing with more than a few MB of data, pexpect is
probably the wrong tool for the job for a variety of reasons.

**Comment by CKD4XXH** [comment]
If the communication speed is not very high and known and near constant then with the timeout parameter I can control the amount of bytes aggregated (memory consumed). But.. perhaps not always is this the case. And an optional **size limit** (would not hurt and) could help.

**Comment by ptsneves** [comment]
I also agree with @CKD4XXH and this struck me as a bit tricky. Perhaps adding some piece of documentation. If an option is possible I can add it.

**Comment by Red-M** [comment]
I'd ask that you check with the original tcl expect to see if this is possible by doing the same actions from pexpect. I doubt that this is an implementation issue.
